src/provider/meta.ts

```diff
@@ -0,0 +1,63 @@
+import type { ExternalProvider, JsonRpcProvider, Web3Provider } from '@ethersproject/providers'
+import WalletConnectProvider from '@walletconnect/ethereum-provider'
+
+function isWeb3Provider(provider: JsonRpcProvider): provider is Web3Provider {
+  return 'provider' in provider
+}
+
+function isWalletConnectProvider(provider: ExternalProvider): provider is WalletConnectProvider {
+  return (provider as WalletConnectProvider).isWalletConnect
+}
+
+export enum WalletType {
+  WALLET_CONNECT = 'WalletConnect',
+  INJECTED = 'Injected',
+  UNKNOWN = 'Unknown',
+}
+
+export interface WalletMeta extends Partial<NonNullable<WalletConnectProvider['connector']['peerMeta']>> {
+  type: WalletType
+}
+
+function getWalletConnectMeta(provider: WalletConnectProvider): WalletMeta {
+  return {
+    type: WalletType.WALLET_CONNECT,
+    ...provider.connector.peerMeta,
+  }
+}
+
+function getInjectedMeta(provider: ExternalProvider & Record<string, unknown>): WalletMeta {
+  const properties = Object.getOwnPropertyNames(provider)
+  const names =
+    properties
+      .filter((name) => name.match(/^is.*$/) && (provider as Record<string, unknown>)[name] === true)
+      .map((name) => name.slice(2)) ?? []
+
+  // Add qrUrl to identify QR-initiated connections (eg for Coinbase Wallet via mobile QR).
+  if (properties.includes('qrUrl') && provider['qrUrl']) {
+    names.push('qrUrl')
+  }
+
+  // Sort MetaMask last, so that wallets spoofing MetaMask list themselves first.
+  names.sort((a, b) => (a === 'MetaMask' ? 1 : b === 'MetaMask' ? -1 : 0))
```

Comment:
In that case, why would MetaMask be still present in an array? In theory, I would expect this to just return a single wallet.

Otherwise, `names.join()` will be like `Coinbase MetaMask` which I don't think is true? (you're not connected with two wallets via a single provider)

Reply:
I'm creating a new property called `agent` which will have all of the strings in the array (like a user-agent), and will just return the first word for `name`.

Reply:
Sounds good, that could also let us know whether it's an injected provider or not (since most of them advertises as Metamask anyway for compatibility).